Feature: knownly deployment provider

dwightgunning opened the pull request:
Knownly is a service that allows people to host websites from a Dropbox folder. This PR allows people to deploy to their Knownly website folder - a local Dropbox folder (or any other folder in fact).

The deployment provider functions in a similar manner to the other existing providers such as S3 and Rackspace. First the developer just adds the custom provider to the config.json. The first 'deploy' will prompt for a target destination which is saved in the config for future.

krallin commented on cactus/deployment/knownly/engine.py:
```diff
@@ -0,0 +1,75 @@
+        if not os.path.exists(self.target_directory):
+            logger.info("Knownly website folder does not exist. Cancelling " \
+                        "deployment.")
+            return []
```
Note to self: don't we want to throw an exception itself.

krallin commented on cactus/deployment/knownly/engine.py:
```diff
@@ -0,0 +1,75 @@
+        # Upload all files concurrently in a thread pool
+        mapper = multiMap
+        if self.site._parallel <= PARALLEL_DISABLED:
+            mapper = map
```
Note to self: need to review whether this will need work with the Python 3 support migration.

krallin commented:
Hey there, and thanks for your contributions!

I had a few notes for you,

+ First one is about the use of `filecmp`. According to the Python docs, it only looks at `os.stat`, which might not be the most reliable way to identify if two files are in fact identical. What do you think? 
+ Second is about tests. Did you look into whether it was reasonable feasible to add some here? (intuitively, I think that's the case) I can also look into it at some point (but I've been badly backlogged as of late).
+ Third is about the Knownly name. There seems to be little that is specific to Knownly here, am I missing something?

If my understanding is correct that there is nothing specific to Knownly, how about:

+ Remove the Knownly name from the code and use something a bit more explicit (e.g. folder or file, etc.)
+ Use knownly as the name of the deployment engine that people use in their config (and possibly also use a more generic name pointing to the same deployment engine).

This way, the internals use something explicit, and the user-visible parts use the Knownly name;.

Note: I do understand and respect the fact that you might want to promote Knownly as a first-party integration. I also think it's a fair ask considering you're providing something that has definitely been requested before by our user base, but I'm hoping we can find a compromise that'll work well! I'm definitely open to other ideas as well.

Cheers,

dwightgunning commented:
Hi,

Thanks for the kind welcome and feedback.

Let me address your initial notes:
 - I chose `filecmp` as I think it should be reliable in majority of cases. At a minimum it looks at file size and modification dates. I could be wrong but I doubt the other providers (S3, Rackspace) do full binary diffs so this seemed a fair trade-off for speed.
 - I can attempt writing some unit tests for this new provider.
 - You're correct there isn't anything truly Knownly specific here (+1 for Knownly over the other deployment destinations ;)

I'm fine to re-write and re-raise the PR with more generic naming in the source. I like your suggestion to use 'knownly' as the config label and perhaps we can reflect the contribution in the docs?

Is there any timeline for a next release that I should work toward?

krallin commented:
Hey there,

+ If I recall correctly, the S3 provider (I don't remember about the others off the top of my head) actually checks the checksum (which is provided by S3). Not sure about the others, but I could check.
+ Thanks for looking into tests.

Feel free to make updates to this PR directly.  There is no specific timeline for the next release, but considering this is simply a new feature with little potential for breakage, we can ship it when it's ready (note that I'm speaking about the CLI tool here, not the Mac App). 

Thanks!

dwightgunning commented:
Hi @krallin I finally got around to re-writing the filesystem provider.

Removed references to my Knownly service, rebased the feature-branch on master, should be good to go.

dwightgunning commented:
Housekeeping close.
